source/dub/dependencyresolver.d

```diff
@@ -54,236 +54,149 @@ class DependencyResolver(CONFIGS, CONFIG) {
 		}
 	}
 
-	private static struct PackageConfigs
-	{
-		static struct Depender
-		{
-			TreeNode origin;
-			TreeNodes dependency;
-		}
-
-		// all possible configurations to test for this package
-		CONFIG[] allConfigs;
-
-		// determines whether this package has any dependencies, may be
-		// different from allConfigs.length > 0 after certain configurations
-		// have been filtered out
-		bool anyConfig;
-
-		Depender[] origins;
-	}
-
 	CONFIG[string] resolve(TreeNode root, bool throw_on_failure = true)
 	{
-		auto root_base_pack = basePackage(root.pack);
+		import std.bitmanip : BitArray;
 
-		// find all possible configurations of each possible dependency
-		size_t[string] package_indices;
-		string[size_t] package_names;
-		PackageConfigs[] configs;
-		bool[string] maybe_optional_deps;
-		bool[TreeNode] visited;
+		// Leave the possibility to opt-out from the loop limit
+		import std.process : environment;
+		bool no_loop_limit = environment.get("DUB_NO_RESOLVE_LIMIT") !is null;
 
-		void findConfigsRec(TreeNode parent, bool parent_unique)
-		{
-			if (parent in visited) return;
-			visited[parent] = true;
-
-			foreach (ch; getChildren(parent)) {
-				auto basepack = basePackage(ch.pack);
-				auto pidx = configs.length;
-
-				if (ch.depType != DependencyType.required) maybe_optional_deps[ch.pack] = true;
-
-				PackageConfigs config;
-				if (auto pi = basepack in package_indices) {
-					pidx = *pi;
-					config = configs[*pi];
-				} else {
-					if (basepack == root_base_pack) config.allConfigs = [root.config];
-					else config.allConfigs = getAllConfigs(basepack);
-					configs ~= config;
-					package_indices[basepack] = pidx;
-					package_names[pidx] = basepack;
+		// build up the dependency graph, eliminating as many configurations/
+		// versions as possible
+		PackageConfigs[string] packs;
```

Comment:
should we maybe call this `cache` or `visited` instead to make more clear what packs are stored in this variable?